applications/nrf_desktop/src/modules/ble_qos.c

```diff
@@ -0,0 +1,741 @@
+	struct params_wifi wifi_params;
+	struct chmap_filter_params *params;
+	size_t pos;
+
+	params = k_malloc(sizeof(struct chmap_filter_params));
+	if (params == NULL) {
+		LOG_ERR("OOM error");
+		return;
+	}
+
+	chmap_filter_params_get(chmap_inst, params);
+
+	if (qos_ble_params != NULL) {
+		pos = 0;
+
+		params->maintenance_sample_count =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.sample_count_min);
+
+		params->min_channel_count = qos_ble_params[pos];
+		pos += sizeof(ble_params.min_channel_count);
+
+		params->ble_weight_crc_ok =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.weight_crc_ok);
+
+		params->ble_weight_crc_error =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.weight_crc_error);
+
+		params->ble_block_threshold =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.ble_block_threshold);
+
+		params->eval_max_count = qos_ble_params[pos];
+		pos += sizeof(ble_params.eval_max_count);
+
+		params->eval_duration =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.eval_duration);
+
+		params->eval_keepout_duration =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.eval_keepout_duration);
+
+		params->eval_success_threshold =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(ble_params.eval_success_threshold);
+	}
+	if (qos_wifi_params != NULL) {
+		pos = 0;
+
+		params->wifi_rating_inc =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(wifi_params.wifi_rating_inc);
+
+		params->wifi_present_threshold =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(wifi_params.wifi_present_threshold);
+
+		params->wifi_active_threshold =
+			sys_get_le16(&qos_ble_params[pos]);
+		pos += sizeof(wifi_params.wifi_active_threshold);
+	}
+
+	k_mutex_lock(&data_access_mutex, K_FOREVER);
+	if (new_params != NULL) {
+		k_free(new_params);
+	}
+	new_params = params;
+	k_mutex_unlock(&data_access_mutex);
+}
+
+static void update_blacklist(const u8_t *blacklist)
+{
+	atomic_set(&new_blacklist, sys_get_le16(blacklist));
+}
+
+static int hci_chmap_cmd_send(const u8_t *chmap)
+{
+	struct bt_hci_cp_le_set_host_chan_classif *cp;
+	struct net_buf *buf;
+
+	__ASSERT_NO_MSG(chmap != NULL);
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_HOST_CHAN_CLASSIF,
+				sizeof(*cp));
+	if (!buf) {
+		return -1;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	if (!cp) {
+		return -1;
+	}
+
+	memcpy(cp->ch_map, chmap, sizeof(cp->ch_map));
+
+	return bt_hci_cmd_send_sync(
+			BT_HCI_OP_LE_SET_HOST_CHAN_CLASSIF,
+			buf,
+			NULL);
+}
+
+static bool on_vs_evt(struct net_buf_simple *buf)
+{
+	u8_t *subevent_code;
+	hci_vs_evt_qos_conn_event_report_t *evt;
+
+	subevent_code = net_buf_simple_pull_mem(
+						buf,
+						sizeof(*subevent_code));
+
+	switch (*subevent_code) {
+	case HCI_VS_SUBEVENT_CODE_QOS_CONN_EVENT_REPORT:
+		if (atomic_get(&processing) == true) {
+			/* Cheaper to skip this update */
+			/* instead of using locks */
```

Comment:
Yes, that's the design. Adding a build assert to enforce the priority ordering

Reply:
ok, but this is a function that anyone may call from wrong context in the future. Where is the design constraint provided for people that may change the code?

Reply:
Currently one have to read the function documentation to understand the context restriction. The library could be updated to use atomic data types to avoid this restriction, at the expense of increased memory usage due to inflexible size of atomic data type